relayer/strategies.go

```diff
@@ -32,62 +31,83 @@ func relayerMainLoop(ctx context.Context, log *zap.Logger, src, dst *Chain, filt
 		// So we will occasionally query recent txs and check the events for `ChannelOpenInit`, at which point
 		// we will attempt to finish opening the channel.
 
+		// TODO when we have completed the Chain/Path processor refactor we will be listening for
+		// new channels that are being opened/created so it's possible we have no open channels to relay on
+		// at startup but after some time has passed a channel needs opened and relayed on. At this point we
+		// could choose to loop here until some action is needed.
 		if len(srcOpenChannels) == 0 {
-			continue
+			errCh <- fmt.Errorf("there are no open channels to relay on, check your config's allow/deny list rules")
+			return
 		}
 
-		var wg sync.WaitGroup
-		// Spin up a goroutine to relay packets & acks for each channel that isn't already being relayed against
+		// Spin up a goroutine to relay packets & acks for each channel that isn't already being relayed against.
 		for _, channel := range srcOpenChannels {
 			if !channel.active {
 				channel.active = true
-				wg.Add(1)
-				go relayUnrelayedPacketsAndAcks(ctx, log, &wg, src, dst, maxTxSize, maxMsgLength, channel, channels)
+				go relayUnrelayedPacketsAndAcks(ctx, log, src, dst, maxTxSize, maxMsgLength, channel, channels)
 			}
 		}
-		wg.Wait()
 
-		for channel := range channels {
-			channel.active = false
-			break
+		// Block here until one of the running goroutines exit.
+		channel, ok := <-channels
+		if !ok {
+			errCh <- fmt.Errorf("the (chan *ActiveChannel) was closed and no more values can be written/read")
+			return
 		}
 
-		// Make sure we are removing channels no longer in OPEN state from the slice of open channels
-		for i, channel := range srcOpenChannels {
-			if channel.channel.State != types.OPEN {
-				srcOpenChannels[i] = srcOpenChannels[len(srcOpenChannels)-1]
-				srcOpenChannels = srcOpenChannels[:len(srcOpenChannels)-1]
+		channel.active = false
+
+		// When a goroutine exits we need to query the channel and check that it is still in OPEN state.
+		var queryChannelResp *types.QueryChannelResponse
+		if err = retry.Do(func() error {
+			queryChannelResp, err = src.ChainProvider.QueryChannel(ctx, 0, channel.channel.ChannelId, channel.channel.PortId)
+			if err != nil {
+				return err
 			}
+			return nil
+		}, retry.Context(ctx), RtyAtt, RtyDel, RtyErr, retry.OnRetry(func(n uint, err error) {
+			src.log.Info(
+				"Failed to query channel for updated state",
+				zap.String("src_chain_id", src.ChainID()),
+				zap.String("src_channel_id", channel.channel.ChannelId),
+				zap.Uint("attempt", n+1),
+				zap.Uint("max_attempts", RtyAttNum),
+				zap.Error(err),
+			)
+		})); err != nil {
+			errCh <- err
+			return
+		}
+
+		// If the channel is no longer in OPEN state then we remove it from the map of open channels.
+		if queryChannelResp.Channel.State != types.OPEN {
+			delete(srcOpenChannels, channel.channel.ChannelId)
```

Comment:
Seems worth logging, if a channel is no longer open.

Then, do we need a special case to check if `len(srcOpenChannels) == 0` again? Right now it will hit the check at the beginning of the for-loop, which isn't exactly accurate if a channel used to be open but now isn't.

Reply:
I agree, logging here is nice.

Given that sometime in the near future we will likely still loop even if there are no channels open, I think I'd rather revise the error message near the top instead of adding another check on the number of open channels.